src/bgem/bspline/bspline_approx.py

```diff
@@ -956,51 +809,74 @@ def _init_coo_structure(self):
+
+    def _locate_points(self):
         """
         Construction of the system B^TWBz=B^TWb
         for control points of the 2th order B-spline surface
         """
-        normal_matrix_size = self._u_basis.size * self._v_basis.size
+        point_loc = []
         n_points = self._uv_quad_points.shape[0]
-        row, col, data, point_loc = self._init_coo_structure()
-        vec_BTb = np.zeros(normal_matrix_size)
-        avg_vec = np.ones(normal_matrix_size)
-        avg = 0
+        
+        for i in range(0,(self._u_basis.size - 2) *(self._v_basis.size - 2)):
+            point_loc.append([])
 
         for idx in range(n_points):
             u, v = self._uv_quad_points[idx, 0:2]
-            b_entry = self._z_quad_points[idx]
-            avg += b_entry
             iu = self._u_basis.find_knot_interval(u)
             iv = self._v_basis.find_knot_interval(v)
             idp = self.patch_pos2id(iu, iv)
-            u_base_vec = self._u_basis.eval_vector(iu, u)
-            v_base_vec = self._v_basis.eval_vector(iv, v)
-            data9 = np.kron(v_base_vec, u_base_vec)
-            b_row = col[idp * 81: idp * 81 + 9]
-            ##  B^TWBz=B^TWb
-            w_data9 = self._w_quad_points[idx] * data9
-            data[idp*81:(idp+1)*81] += np.kron(data9, w_data9)
-            vec_BTb[b_row.tolist()] += b_entry * w_data9
+            u_base_vec = self._u_basis.eval_vector(iu, u).transpose()
+            v_base_vec = self._v_basis.eval_vector(iv, v).transpose()
             point_loc[idp].append(idx)
 
-        avg = avg / n_points
-        avg_vec = avg_vec * avg
+        self.point_loc = point_loc
 
+    def _build_system_of_normal_equations(self):
+        """
+        Construction of the system B^TWBz=B^TWb
+        for control points of the 2th order B-spline surface
+        """
+        normal_matrix_size = self._u_basis.size * self._v_basis.size
+        n_patches = (self._u_basis.size - 2) * (self._v_basis.size - 2)
+        n_points = self._uv_quad_points.shape[0]
+        row, col, data = self._init_coo_structure()
+        vec_BTb = np.zeros(normal_matrix_size)
+        avg_vec = np.ones(normal_matrix_size) * np.sum(self._z_quad_points)/n_points
+
+        for patch_id in range(0, n_patches):
+            if len(self.point_loc[patch_id]) > 0:
+                patch_point_loc = self.point_loc[patch_id]
+                u_loc_vec = self._uv_quad_points[patch_point_loc, 0]
+                v_loc_vec = self._uv_quad_points[patch_point_loc, 1]
+                w_loc_vec = self._w_quad_points[patch_point_loc]
+                b_loc_vec = self._z_quad_points[patch_point_loc]
+                iu, iv = self.patch_id2pos(patch_id)
+                u_loc_base_vec = self._u_basis.eval_vector(iu, u_loc_vec).transpose()
+                v_loc_base_vec = self._v_basis.eval_vector(iv, v_loc_vec).transpose()
+                v_kron_u = (v_loc_base_vec[:, :, None] @ u_loc_base_vec[:, None, :]).reshape(len(self.point_loc[patch_id]),9)
```

Comment:
Why matrix mult?
Rather use expressions with short axes first, that can use vector operations and on most modern processors that leads to 4 or 8 times faster execution.

E.g.
```
#!/usr/bin/env python
# coding: utf-8

# In[2]:


import numpy as np
import timeit


# In[5]:



n=10000
points_3n = np.random.rand(3, n)
points_n3 = points_3n.transpose().copy()


# In[11]:


def operation_3n(points):
    n = points.shape[1]
    w = points[0, :]
    loc_matrix = (points[None,:,:] * points[:,None,:]).reshape(9, n)
    w_loc_mat = w[None, :n] * loc_matrix[:9, :n]
    res = w_loc_mat[:9,None, :n] * w_loc_mat[None, :9, :n]
    
    

def operation_n3(points):
    n = points.shape[0]
    w = points[:, 0]
    loc_matrix = (points[:, None,:] * points[:, :, None]).reshape(n, 9)
    w_loc_mat = w[:n, None] * loc_matrix[:n, :9]
    res = w_loc_mat[:n, :9,None] * w_loc_mat[:n, None, :9]

print(timeit.timeit(lambda: operation_3n(points_3n), number=10000))
print(timeit.timeit(lambda: operation_n3(points_n3), number=10000)) 
# Output:
# 5.025110515998676
# 21.568570670002373

```